imap_processing/lo/l0/science_direct_events.py

```diff
@@ -76,45 +65,54 @@ class ScienceDirectEvents(LoBase):
 
     SHCOARSE: int
     COUNT: int
-    DATA: bitstring.Bits
+    DATA: str
     CHKSUM: int
-    TOF0: float
-    TOF1: float
-    TOF2: float
-    TOF3: float
-    TIME: float
-    ENERGY: float
-    POS: float
+    TOF0: np.ndarray
+    TOF1: np.ndarray
+    TOF2: np.ndarray
+    TOF3: np.ndarray
+    TIME: np.ndarray
+    ENERGY: np.ndarray
+    POS: np.ndarray
 
     def __init__(self, packet, software_version: str, packet_file_name: str):
         """Intialization method for Science Direct Events Data class."""
         super().__init__(software_version, packet_file_name, CcsdsData(packet.header))
         self.parse_data(packet)
+        self.TOF0 = np.array([])
+        self.TOF1 = np.array([])
+        self.TOF2 = np.array([])
+        self.TOF3 = np.array([])
+        self.TIME = np.array([])
+        self.ENERGY = np.array([])
+        self.POS = np.array([])
         self._decompress_data()
 
     def _decompress_data(self):
         """Decompress the Lo Science Direct Events data."""
-        case_number = self._find_decompression_case()
-        tof_decoder = self._find_tof_decoder_for_case(case_number)
-        tof_calculation_binary = self._read_tof_calculation_table(case_number)
-        remaining_bit_coefficients = self._find_remaining_bit_coefficients(
-            tof_calculation_binary
-        )
-        parsed_bits = self._parse_binary(case_number, tof_decoder)
-        # decode_fields will set the TOF class variables
-        self._decode_fields(remaining_bit_coefficients, parsed_bits)
+        bitstream = bitstring.ConstBitStream(bin=self.DATA)
+        for _ in self.COUNT:
+            case_number = self._find_decompression_case(bitstream)
+            tof_decoder = self._find_tof_decoder_for_case(case_number, bitstream)
+            tof_calculation_binary = self._read_tof_calculation_table(case_number)
+            remaining_bit_coefficients = self._find_remaining_bit_coefficients(
+                tof_calculation_binary
+            )
+            parsed_bits = self._parse_binary(case_number, tof_decoder, bitstream)
+            # decode_fields will set the TOF class variables
+            self._decode_fields(remaining_bit_coefficients, parsed_bits)
 
-    def _find_decompression_case(self):
+    def _find_decompression_case(self, bitstream):
         """Find the decompression case for this DE.
 
         The first 4 bits of the binary data are used to
         determine which case number we are working with.
         The case number is used to determine how to
         decompress the TOF values.
         """
-        return int(self.DATA[0:4].bin, 2)
+        return bitstream.read(4).uint
```

Comment:
If you pass the current location in, you could keep this as it was before I think.
```suggestion
        return int(self.DATA[loc:loc+4], 2)
```